domain/resourcegroup/runaway.go

```diff
@@ -75,84 +79,291 @@ type RunawayRecord struct {
-	key := resourceGroupName + "/" + convict
-	if rm.watchList.Get(key) == nil {
-		rm.queryLock.Lock()
-		if rm.watchList.Get(key) == nil {
-			rm.watchList.Set(key, struct{}{}, ttl)
-		}
-		rm.queryLock.Unlock()
-	}
+func (rm *RunawayManager) markQuarantine(resourceGroupName, convict string, watchType rmpb.RunawayWatchType, action rmpb.RunawayAction, ttl time.Duration, now *time.Time) {
 	var endTime time.Time
-	if ttl == 0 {
-		endTime = longestTime
-	} else {
-		endTime = now.Add(ttl)
+	if ttl > 0 {
+		endTime = now.UTC().Add(ttl)
 	}
-	select {
-	case rm.quarantineChan <- &QuarantineRecord{
+	record := &QuarantineRecord{
 		ResourceGroupName: resourceGroupName,
-		StartTime:         *now,
+		StartTime:         now.UTC(),
 		EndTime:           endTime,
 		Watch:             watchType,
 		WatchText:         convict,
 		Source:            rm.serverID,
-	}:
+		Action:            action,
+	}
+	rm.addWatchList(record, ttl, false)
+	select {
+	case rm.quarantineChan <- record:
 	default:
 		// TODO: add warning for discard flush records
 	}
 }
+
+func (rm *RunawayManager) addWatchList(record *QuarantineRecord, ttl time.Duration, force bool) {
+	key := record.GetRecordKey()
+	// This is a pre-check, because we generally believe that in most cases, we will not add a watch list to a key repeatedly.
+	item := rm.getWatchFromWatchList(key)
+	if force {
+		rm.queryLock.Lock()
+		defer rm.queryLock.Unlock()
+		if item != nil {
+			if item.ID == record.ID {
+				return
+			}
+			rm.watchList.Delete(key)
+		}
+		rm.watchList.Set(key, record, ttl)
+	} else {
+		if item == nil {
+			rm.queryLock.Lock()
+			// When watchlist get record, it will check whether the record is stale, so add new record if returns nil.
+			if rm.watchList.Get(key) == nil {
+				rm.watchList.Set(key, record, ttl)
+			} else {
+				rm.staleQuarantineRecord <- record
+			}
+			rm.queryLock.Unlock()
+		} else if item.ID != record.ID {
+			rm.staleQuarantineRecord <- record
+		}
+	}
+}
+
+// GetWatchByKey is used to get a watch item by given key.
+func (rm *RunawayManager) GetWatchByKey(key string) *QuarantineRecord {
+	return rm.getWatchFromWatchList(key)
+}
+
+// GetWatchList is used to get all watch items.
+func (rm *RunawayManager) GetWatchList() []*QuarantineRecord {
+	items := rm.watchList.Items()
+	ret := make([]*QuarantineRecord, 0, len(items))
+	for _, item := range items {
+		ret = append(ret, item.Value())
+	}
+	return ret
+}
+
+// AddWatch is used to add watch items from system table.
+func (rm *RunawayManager) AddWatch(record *QuarantineRecord) {
+	ttl := time.Until(record.EndTime)
+	if record.EndTime.Equal(NullTime) {
+		ttl = 0
+	} else if ttl <= 0 {
+		rm.staleQuarantineRecord <- record
+		return
+	}
+
+	force := false
+	// The manual record replaces the old record.
+	force = record.Source == ManualSource || record.Source == rm.serverID
+	rm.addWatchList(record, ttl, force)
+}
+
+// RemoveWatch is used to remove watch items from system table.
```

Comment:
misleading comment, this function only remove records from in-memory cache

Reply:
Well, it does have ambiguity